sklearn/preprocessing/tests/test_imputation.py

```diff
@@ -357,3 +362,97 @@ def test_imputation_copy():
+             [11,  -1,   1,  1]
+    ])
+
+    def assert_type(actual, is_sparse, sp, missing_values):
+        if sp is True:
+            assert_equal(actual, sparse.csc_matrix)
+        elif ((sp is "auto" and missing_values == 0) or sp is False):
+            assert_equal(actual, np.ndarray)
+        else:
+            if is_sparse:
+                assert_equal(actual, sparse.csc_matrix)
+            else:
+                assert_equal(actual, np.ndarray)
+
+    def assert_mask(actual, expected, features):
+        if hasattr(actual, 'toarray'):
+            assert_array_equal(actual.toarray(), expected[:, features])
+        else:
+            assert_array_equal(actual, expected[:, features])
+
+    def _check_missing_indicator(X1, X2, retype, sp, missing_values):
+        mask = X2 == missing_values
+        expect_feat_missing = np.where(np.any(X1 == missing_values, axis=0))[0]
+
+        X1_in = retype(X1)
+        X2_in = retype(X2)
+        # features = "train":
+        MI = MissingIndicator(missing_values=missing_values,
+                              sparse=sp)
+
+        MI.fit(X1_in)
+        X2_tr = MI.transform(X2_in)
+        features = MI.feat_with_missing_
+        assert_array_equal(expect_feat_missing, features)
+        assert_type(type(X2_tr), sparse.issparse(X2_in), sp, missing_values)
+        assert_mask(X2_tr, mask, features)
+
+        # features = "all"
+        MI = clone(MI).set_params(features="all")
+        MI.fit(X1_in)
+        X2_tr = MI.transform(X2_in)
+        features = np.arange(X2.shape[1])
+        assert_mask(X2_tr, mask, features)
+
+        # features = [1, 2]
+        features = [1, 2]
+        MI = clone(MI).set_params(features=features)
+        MI.fit(X1_in)
+        X2_tr = MI.transform(X2_in)
+        assert_mask(X2_tr, mask, features)
+
+    for X1, X2, missing_values in [(X1_orig, X2_orig, -1),
+                                   (X1_orig + 1, X2_orig + 1, 0)]:
+        for retype in [lambda x: x.tolist(), np.array, sparse.csr_matrix,
+                       sparse.csc_matrix, sparse.lil_matrix]:
+            for sp in [True, False, 'auto']:
+                _check_missing_indicator(X1, X2, retype, sp, missing_values)
+
+
+def test_missing_indicator_warning():
+    X1 = np.array([
+          [-1,  1,  3],
+          [4,  0, -1],
+          [8,  1,  0]
+    ])
+    X2 = np.array([
+          [5, -1, -1],
+          [-1,  2,  3],
+          [2,  4,  0]
+    ])
+    MI = MissingIndicator(missing_values=-1)
+    MI.fit(X1)
+    missing_features_fit = np.where(np.any(X1 == -1, axis=0))[0]
+    missing_features_tr = np.where(np.any(X2 == -1, axis=0))[0]
+    extra_missing_features = np.setdiff1d(missing_features_tr,
+                                          missing_features_fit)
+    warn_msg = "The features %s have missing values in transform " \
+               "but have no missing values in fit" % extra_missing_features
+    assert_warns_message(RuntimeWarning, warn_msg, MI.transform, X2)
```

Comment:
we also need to check that the same warning is not raised for `features='all'` or `features=[0, 2]`